core/src/connection.rs

```diff
@@ -80,6 +84,9 @@ pub enum ConnectedPoint {
         ///   connection as a dialer and one peer dial the other and upgrade the
         ///   connection _as a listener_ overriding its role.
         role_override: Endpoint,
+        /// Whether the port for the outgoing connection was reused from a listener
+        /// or a new port was allocated. This is useful for address translation.
+        port_use: PortUse,
```

Comment:
Thanks for flagging.

I guess the question boils down to which of the two paths below we want to take. In case `libp2p-swarm` specifies `PortUse::Reuse` and the bottom layer `Transport` has no listener available, i.e. no listening port it can reuse:

1. Should the `Transport` have some means to report that it was instructed to reuse a port but wasn't able to?
2. Or should we just consider it a best effort and simply put into `ConnectedPoint::Dialer` what `libp2p-swarm` instructed the `Transport` to do?

For (1) we would need to either change the `Transport::dial` return type signature or the `Transport::Dial` associated type. For (2) we don't need to do any changes to either of the two.

If we follow (2) `Map` in `transport/map.rs` would simply set `ConnectedPoint::Dialer::port_use` to whatever it got as an input parameter through `opts.port_use`.

What do you think @thomaseizinger and @umgefahren? Should we start off with (2)?

Reply:
Another possibility would be to split `Map` into `MapDial` and `MapListen`, that way, we don't have to make use of `ConnectedPoint` here.

Alternatively, we can also just split the enum and use a new one for our purposes.

Reply:
I thought (2) is what we agreed upon? For combinators like `transport::Map`, we should definitely just pass it on.

Reply:
Can you explain a little bit more what's supposed to happen in Map and AndThen and so on, with ConnectedPoint? I think there is the issue here for me.

Reply:
> I thought the plan was that we return the information whether a port was reused.

The way I understood it was that `Swarm` passes the `PortUse` variable _down_ to `Transport` and once the connection is established, it passes the same variable on to the `NetworkBehaviour`. In other words, the transport can't report whether it actually use a new or an existing port.

To do that, we'd have to change the signature of `Transport::dial`.

> Can you explain a little bit more what's supposed to happen in Map and AndThen and so on, with ConnectedPoint? I think there is the issue here for me.

Here is how I imagine this:

```rust
    fn dial(&mut self, addr: Multiaddr, opts: DialOpts) -> Result<Self::Dial, TransportError<Self::Error>> {
        let future = self.transport.dial(addr.clone(), opts)?;
        let p = ConnectedPoint::Dialer {
            address: addr,
            role_override: Endpoint::Dialer,
            port_use: opts.port_use
        };
        Ok(MapFuture {
            inner: future,
            args: Some((self.fun.clone(), p)),
        })
    }
```

Does that not work?

Reply:
> I thought (2) is what we agreed upon? For combinators like `transport::Map`, we should definitely just pass it on.

:+1: I wasn't sure anymore. (2) sounds good to me.

> To do that, we'd have to change the return value of `Transport::dial`. Is that worth it?

I don't think it is worth it. Let's drop (1) and go with (2).

Reply:
Ok I will do that

Reply:
I hope you like the changes.